MFC Containerization

Malmahrouqi3 opened the pull request:
### **User description**
## Description
Replacing (#935), since it was overly inefficient workflow without yielding any tangible advantages.

**Notes to Self:**
- MFC Docker Repo Overview (README)
- Docker Desktop GUI to run containers.
<img width="1657" height="364" alt="image" src="https://github.com/user-attachments/assets/ffac2374-ce40-4b93-a267-67eec7621b5d" />

- SLURM job script templates
- Config case (case.py) and run/profile on a cluster.
- Checkout [Resource constraints](https://docs.docker.com/engine/containers/resource_constraints/) if useful.
- Obsoleteness vs. Potential of QEMU to support more systems.
```
    tests/UUID     (s)      Summary
QEMU Emulation
     8F78E60C     46.35    1D -> MHD -> HLLD
     2ADA983F    135.76    2D -> MHD -> HLLD
Native Arch Support
     8F78E60C      3.17    1D -> MHD -> HLLD
     2ADA983F      8.79    2D -> MHD -> HLLD
---------------------------------------------------
    "supported": [
      "linux/amd64",
      "linux/amd64/v2",
      "linux/amd64/v3",
      "linux/arm64",
      "linux/riscv64",
      "linux/ppc64le",
      "linux/s390x",
      "linux/386",
      "linux/mips64le",
      "linux/mips64",
      "linux/loong64",
      "linux/arm/v7",
      "linux/arm/v6"
    ],
    "emulators": [
      "cli",
      "llvm-13-runtime.binfmt",
      "llvm-14-runtime.binfmt",
      "llvm-15-runtime.binfmt",
      "python3.10",
      "qemu-aarch64",
      "qemu-arm",
      "qemu-loongarch64",
      "qemu-mips64",
      "qemu-mips64el",
      "qemu-ppc64le",
      "qemu-riscv64",
      "qemu-s390x"
    ]
```

- Improve User Guide

### **Containerization of MFC Releases (v4.3.0-v5.0.5)**
```
├── .github/
│   ├── Dockerfile           # Contains docker build Instructions
│   ├── .dockerignore        # Contains a list of files and directories to exclude in the docker.
│   └── workflows/
│          └── docker.yml    # Builds and deploys images triggered by MFC release tag input.
│          └── release.yml   # Triggers docker.yml to containerize a range of releases specified by user
```
1. Fork `https://github.com/Malmahrouqi3/MFC-mo2.git --branch docker`, and add credentials to its GitHub secrets.
2. On GitHub Actions, go to Release Dispatcher Workflow (`release.yml`), then hit _Run workflow_ and specify a range of desired releases. Except the first few published releases where `--gpu` flag was not defined yet, the rest should build without any issues. Repeated use of Release Dispatcher for debugging could exhaust docker limits.
e.g. "You have reached your pull rate limit as"
3. On Docker Huh, keep track of pushed releases.

### **Setup MFC Container on Codespaces**
```
├── .devcontainer/
│   └── devcontainer.json    # Contains codespace configuration to load MFC docker container
```
This is intended to pre-configure repo codespace instances to load MFC docker container for users to interact with instantaneously.
1. Create a `.devcontainer` folder in the repo root, and add a `devcontainer.json` file.
```json
{
  "name": "MFC Dev Container",
  "image": "sbryngelson/mfc:<tag>",
  "workspaceFolder": "/opt/MFC",
  "settings": {
    "terminal.integrated.shell.linux": "/bin/bash",
    "editor.formatOnSave": true
  },
}
```


### **Quick Start Guide**
1. **Run an MFC container**

**GitHub Codespaces**
- Start a codespace on the repository and a container will be loaded to interact with.

**Locally**
- Pull from Docker Hub and run the MFC container.
 ```
docker run -it --rm sbryngelson/mfc:v5.0.5-cpu
```
2. **Familiarize yourself with MFC:**
- Run the Test Suite
``` 
./mfc.sh test
```
- Run a simulation case 
```
./mfc.sh run examples/2D_shockbubble/case.py
```

## **User Guide**
Make sure to mount a directory to `mnt` inside the container to easily share files between the host and the container, e.g. `cp -r <source> /mnt/`.

### **Docker CLI**
[Official Documentation](https://docs.docker.com/reference/cli/docker/)
Use it for running and testing containers on your local machine.
**Example Usage:**
```
docker run -it --rm -v "$PWD":/mnt sbryngelson/mfc:<version>-<type: cpu/gpu>
```
**For Portability,**
On the source machine,
- Pull and save the image
```
docker pull sbryngelson/mfc:<tag>
docker save -o mfc:<tag>.tar sbryngelson/mfc:<tag>
```
On the target machine,
- Load and run the image
```
docker load -i mfc:<tag>.tar
docker run -it --rm mfc:<tag>
```

### **Apptainer CLI** 
[Official Documentation](https://apptainer.org/docs/user/main/docker_and_oci.html)
Use it for running containers on powerful machines and HPC clusters on either interactive terminals or batch jobs.
**Example Usage:**
```
apptainer exec --fakeroot --writable-tmpfs --bind "$PWD":/mnt  docker://sbryngelson/mfc:<version>-<type: cpu/gpu> bash -c "cd /opt/MFC && bash"
```
or 
```
apptainer shell --fakeroot --writable-tmpfs --bind "$PWD":/mnt  docker://sbryngelson/mfc:<tag>
Apptainer>cd /opt/MFC
```

`--fakeroot --writable-tmpfs` is critical to:
- Enable root-like permissions inside the container without actual root access
- Allow temporary write access to the container filesystem

**For Portability,**
On the source machine,
- Pull and translate the image into `.sif` format
```
apptainer build mfc:<tag>.sif docker://sbryngelson/mfc:<tag>
```
On the target machine,
- Load and start an interactive shell
```
apptainer shell --fakeroot --writable-tmpfs --bind "$PWD":/mnt mfc:<tag>.sif
```
___

### **PR Type**
Enhancement


___

### **Description**
- Add Docker containerization support for MFC

- Create CPU and GPU container variants

- Implement automated container builds on releases

- Configure Docker Hub deployment workflow


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Source Code"] --> B["Docker Build"]
  B --> C["CPU Container"]
  B --> D["GPU Container"]
  C --> E["Docker Hub"]
  D --> E["Docker Hub"]
  F["GitHub Release"] --> B
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>.dockerignore</strong><dd><code>Docker ignore configuration for clean builds</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/.dockerignore

<ul><li>Add comprehensive Docker ignore patterns<br> <li> Exclude build artifacts, temporary files, and examples output<br> <li> Include specific exceptions for required data files</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-bc25e107942a3f216d54fb99e2047236aa1cdc74adfd210fb29edaa28a1140cc">+83/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Multi-architecture Dockerfile for MFC containers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/Dockerfile

<ul><li>Create multi-stage Dockerfile with CPU/GPU variants<br> <li> Install dependencies based on target architecture<br> <li> Build MFC with appropriate compilers and run tests<br> <li> Set up working environment with proper paths</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-f63117114894cc10c0064b2936899cb8b9e93c2e2d0770cb384e87c37cba9b73">+55/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>docker.yml</strong><dd><code>Automated container build and deployment workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/docker.yml

<ul><li>Add GitHub Actions workflow for container builds<br> <li> Support both release triggers and manual dispatch<br> <li> Build CPU (Ubuntu) and GPU (NVIDIA HPC SDK) variants<br> <li> Deploy containers to Docker Hub with version tags</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-3f5366f6d6df3ec1179e5efadc6f350bfa88eebf4e2da589b4d94ccb85ae5e94">+65/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___


___

### **PR Type**
Enhancement


___

### **Description**
- Add Docker containerization support for MFC releases

- Create GitHub Actions workflow for automated container builds

- Configure development container for Codespaces integration

- Support both CPU and GPU container variants


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["GitHub Release"] --> B["Docker Workflow"]
  B --> C["Build CPU Container"]
  B --> D["Build GPU Container"]
  C --> E["Push to DockerHub"]
  D --> E
  F["Devcontainer Config"] --> G["Codespaces Integration"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>devcontainer.json</strong><dd><code>Add Codespaces development container configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.devcontainer/devcontainer.json

<ul><li>Configure development container for GitHub Codespaces<br> <li> Set MFC Docker image as base container<br> <li> Define workspace folder and editor settings</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-24ad71c8613ddcf6fd23818cb3bb477a1fb6d83af4550b0bad43099813088686">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>.dockerignore</strong><dd><code>Add Docker ignore file for build optimization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/.dockerignore

<ul><li>Exclude build artifacts and temporary files from Docker context<br> <li> Ignore example outputs, logs, and generated files<br> <li> Exclude development tools and cache directories</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-bc25e107942a3f216d54fb99e2047236aa1cdc74adfd210fb29edaa28a1140cc">+82/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Add Dockerfile for MFC containerization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/Dockerfile

<ul><li>Create multi-stage Dockerfile supporting CPU and GPU builds<br> <li> Install dependencies based on target architecture<br> <li> Build MFC with appropriate compiler configurations<br> <li> Set up runtime environment and entrypoint</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-f63117114894cc10c0064b2936899cb8b9e93c2e2d0770cb384e87c37cba9b73">+55/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>docker.yml</strong><dd><code>Add GitHub Actions workflow for container builds</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/docker.yml

<ul><li>Implement automated Docker build workflow<br> <li> Support matrix builds for CPU/GPU and ARM/x64 architectures<br> <li> Create manifest lists for multi-platform support<br> <li> Deploy containers to DockerHub on releases</ul>


</details>


  </td>
  <td><a href="https://github.com/MFlowCode/MFC/pull/971/files#diff-3f5366f6d6df3ec1179e5efadc6f350bfa88eebf4e2da589b4d94ccb85ae5e94">+98/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Adds Dockerfile and CI workflow to build/push CPU/GPU images and a Codespaces devcontainer configuration.
> 
> - **CI/CD**:
>   - `/.github/workflows/docker.yml`: Matrix builds for `cpu`/`gpu` on x86/arm; sets up Buildx/QEMU; clones release tag; builds with compiler/base-image args; pushes to Docker Hub; creates/pushes multi-platform manifest lists (`cpu`, `gpu`, `latest-*`).
> - **Containerization**:
>   - `/.github/Dockerfile`: Parametric base image and compilers; conditional deps for `cpu` vs `gpu`; builds MFC and runs dry-run tests; sets `OMPI_*`, `PATH`, `LD_LIBRARY_PATH`; stages repo in `/opt/MFC` and sets default workdir/entrypoint.
>   - `/.github/.dockerignore`: Excludes build artifacts, caches, example outputs, and large media from build context.
> - **Developer Experience**:
>   - `/.devcontainer/devcontainer.json`: Codespaces devcontainer using published `sbryngelson/mfc:*` image with workspace and editor settings.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bea47cad5bdf13b17e54c511a4ff0912ec58b8f5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Malmahrouqi3 commented:
https://hub.docker.com/r/mohdsaid497566/mfc/tags?ordering=name
I was able to containerize MFC releases down to v4.8.3 without messing with the source code using a [release dispatcher workflow](https://github.com/Malmahrouqi3/MFC-mo2/actions/runs/16609022132/workflow) to trigger docker.yml per MFC release tag. I will open a PR to deploy images for previous MFC releases on `sbryngelson/mfc:tagname` docker repo shortly with instructions for maintainers. I will use them for continuous benchmarking PR (#936).

sbryngelson commented:
Running `docker pull --platform=linux/amd64 mohdsaid497566/mfc:v4.9.6-cpu` grabs on the container on my m1 mac but when it as `docker run --rm -it --platform=linux/amd64 mohdsaid497566/mfc:v4.9.6-cpu` and then `./mfc.sh test` all the tests fail due to segfaults

sbryngelson commented:
Also I no longer see codespaces available in the MFC repo

sbryngelson commented:
for setting up a codespace in the MFC repo (under mflowcode) i have this in my settings:

Codespaces
Prebuild configuration
Codespaces are not enabled for your organization. Please contact your org admin to [enable Codespaces](https://docs.github.com/codespaces/managing-codespaces-for-your-organization/enabling-codespaces-for-your-organization) to use this functionality.
Prebuilds are currently disabled because MFlowCode has reached its Codespaces spending limit for this period. Please contact an administrator to enable prebuilds.

sbryngelson commented:
as someone unfamiliar with docker but familiar with many other "coding thing" I was unable to get this running after trying many different things

sbryngelson commented:
> [hub.docker.com/r/mohdsaid497566/mfc/tags?ordering=name](https://hub.docker.com/r/mohdsaid497566/mfc/tags?ordering=name) I was able to containerize MFC releases down to v4.8.3 without messing with the source code using a [release dispatcher workflow](https://github.com/Malmahrouqi3/MFC-mo2/actions/runs/16609022132/workflow) to trigger docker.yml per MFC release tag. I will open a PR to deploy images for previous MFC releases on `sbryngelson/mfc:tagname` docker repo shortly with instructions for maintainers. I will use them for continuous benchmarking PR (#936).

This sounds like the right path, I just can't run the containers on my machine right now (so new users definitely won't use it in current state. if macos + arm is the issue, well many people own Macbooks...)

Malmahrouqi3 commented:
https://docs.docker.com/build/building/multi-platform/

Found this doc on **multi-platform builds** that _Docker automatically selects the correct variant based on the host's architecture_

sbryngelson commented:
> [docs.docker.com/build/building/multi-platform](https://docs.docker.com/build/building/multi-platform/)
> 
> Found this doc on **multi-platform builds** that _Docker automatically selects the correct variant based on the host's architecture_ The solution is to add to the GitHub Action, `platform: linux/arm64,linux/amd64`

So presumably this builds both versions, I suppose? We could try it out. I feel like this might be the equivalent of me using `docker pull --platform=linux/amd64 mohdsaid497566/mfc:v4.9.6-cpu cpu` which is what i've been doing so far

Malmahrouqi3 commented:
> Oh interesting. I don't see them when doing this on Phoenix nor Bridges2. What computer showed you this, did you have to load other modules first?

Without loading any module, I just tried phoenix, bridges2, and delta with on-demand and ssh, and I was able to use them.

sbryngelson commented:
> > Oh interesting. I don't see them when doing this on Phoenix nor Bridges2. What computer showed you this, did you have to load other modules first?
> 
> Without loading any module, I just tried phoenix, bridges2, and delta with on-demand and ssh, and I was able to use them.

Indeed... apptainer works from compute nodes.

sbryngelson commented:
BTW look at that old MFC logo! very cool

Malmahrouqi3 commented:
**Status Update:**
MFC Docker Repo would look like https://hub.docker.com/r/mohdsaid497566/mfc
where almost all releases are populated along with an overview drafted with decent instructions.

Pre-configured Repo Codespaces https://github.com/Malmahrouqi3/MFC-mo2/
If you start a codespace on my fork, it will load up mohdsaid497566/mfc:v4.4.1-cpu to interact with immediately with no user intervention whatsoever.

Docker GUI can be used to run any containerized release. It would prior stop instantly after startup unless there was an active running command pre-written. However, currently, it starts and stops when instructed by user via GUI buttons. It is pretty cool right now that you can retrieve any release and play with it within seconds under Exec tab.

Malmahrouqi3 commented:
I just finalized the PR and it can be merged safely after populating releases' images (`v4.3.0`-`v5.0.5`) - 38 which should take about a day or so. Not much ought to be changed in the source code in old releases as a result, except for replacing the old SILO block in `CMakeLists.txt` and specifying the version of `cantera==3.0.1` in `pyproject.toml`.

I added `mfc:latest-xxx` tag and simplified a bit the docker readme
[README.txt](https://github.com/user-attachments/files/21649355/README.txt)
[docker repo](https://hub.docker.com/r/mohdsaid497566/mfc)

The typo was introduced 8 months ago.
```bash
mfc: OK > (venv) Entered the Python 3.12.3 virtual environment (>= 3.9).
mfc: Running typos on MFC's toolchain...
error: `frqeuency` should be `frequency`
  --> ./src/common/m_constants.fpp:36:138
   |
36 |     real(wp), parameter :: broadband_spectral_level_growth_rate = 10._wp !< The spectral level constant to correct the magnitude at each frqeuency to ensure the source is overall broadband
   |                                                                                                                                          ^^^^^^^^^
   |
mfc: ERROR > Typos found.
```

sbryngelson commented:
> I just finalized the PR and it can be merged safely after populating releases' images (`v4.3.0`-`v5.0.5`) - 38 which should take about a day or so. Not much ought to be changed in the source code in old releases as a result, except for replacing the old SILO block in `CMakeLists.txt` and specifying the version of `cantera==3.0.1` in `pyproject.toml`.
> 
> I added `mfc:latest-xxx` tag and simplified a bit the docker readme [README.txt](https://github.com/user-attachments/files/21649355/README.txt) [docker repo](https://hub.docker.com/r/mohdsaid497566/mfc)

I'm not quite sure what this means, are you asking someone else to do something, or waiting for some automated process to finish?

Malmahrouqi3 commented:
> > I just finalized the PR and it can be merged safely after populating releases' images (`v4.3.0`-`v5.0.5`) - 38 which should take about a day or so. Not much ought to be changed in the source code in old releases as a result, except for replacing the old SILO block in `CMakeLists.txt` and specifying the version of `cantera==3.0.1` in `pyproject.toml`.
> > I added `mfc:latest-xxx` tag and simplified a bit the docker readme [README.txt](https://github.com/user-attachments/files/21649355/README.txt) [docker repo](https://hub.docker.com/r/mohdsaid497566/mfc)
> 
> I'm not quite sure what this means, are you asking someone else to do something, or waiting for some automated process to finish?

My bad. Since I do not have access to your docker credentials, can you please do it so they all be posted in under `sbryngelson/mfc`?

sbryngelson commented:
> Or to make it simpler can you add me as a collaborator in the docker repo `sbryngelson/mfc` so I would push all stuff in there? Username: `mohdsaid497566`

It seems like we should make a username `mflowcode` or something? I don't want to hold sole responsibility for maintaining this.

Malmahrouqi3 commented:
I have been pushing stuff but it is taking some time
https://hub.docker.com/r/sbryngelson/mfc/tags

sbryngelson commented:
Oh wow. We don't have to go forever back, but doing by release for some sensible amount of time makes sense. What is the limit to how far back we can go in terms of (a) logistics [things like breaking changes to the ./mfc.sh command] and (b) size of dockerhub space? also, does this autopopulate new tags/releases?

Malmahrouqi3 commented:
> Oh wow. We don't have to go forever back, but doing by release for some sensible amount of time makes sense. What is the limit to how far back we can go in terms of (a) logistics [things like breaking changes to the ./mfc.sh command] and (b) size of dockerhub space? also, does this autopopulate new tags/releases?

As far as I am concerned no storage limitation for docker images sorta similar to GitHub. The workflow file `docker.yml` would be triggered by a release or a manual input in case the build failed for whatever reason so no overhead to carry on in terms of containerization overall. The workflow I am using for all previous releases is slightly different but does the same job as the one in this PR.

sbryngelson commented:
status? it's unclear to me if this is 'working', to what degree it has been tested, how it works, how i should test it, etc.?

sbryngelson commented:
for this part 

`Please refer to the docker repository readme for more details on how to properly use containers.`

can we just put the docker repository readme in the main docs as a separate tab under documentation? that way it's all in one place. in the docker repo you can put `Please see the MFC documentation on Docker use [here](link) for more details on its use`

Also we should include documentation on how people can use dev containers from the root repo under clone. That isn't obvious to people that it even exists. they should also be documentation on which container to pick based on what computer you're using.

Malmahrouqi3 commented:
> for this part
> 
> `Please refer to the docker repository readme for more details on how to properly use containers.`
> 
> can we just put the docker repository readme in the main docs as a separate tab under documentation? that way it's all in one place. in the docker repo you can put `Please see the MFC documentation on Docker use [here](link) for more details on its use`
> 
> Also we should include documentation on how people can use dev containers from the root repo under clone. That isn't obvious to people that it even exists. they should also be documentation on which container to pick based on what computer you're using.

Sure thing, I will push all docs to the PR tonight.
For platform/arch, docker picks it by default but yeah I will make sure to indicate that.
For dev containers, I thought of a single line like "start a GitHub Codespace to load a MFC container to interact with" but pointing to how to start one and what to expect is needed as per Dan recom or more dummy-proof wording I guess would be better.

Other than that, no changes to be made to the PR code, just existing MFC docs & README.

Malmahrouqi3 commented:
> yeah all of this should be documented. though the M-series mac chips should work out of the box since so many people run those. there should be M-series github runners for free, no?

List of GitHub runner https://docs.github.com/en/actions/reference/runners/github-hosted-runners
### Linux	          
x64	ubuntu-latest, ubuntu-24.04, ubuntu-22.04
### Windows           
x64	windows-latest, windows-2025, windows-2022
### macOS              
Intel	macos-13, macos-15-intel
### macOS              
arm64	macos-latest, macos-14, macos-15, macos-26 (public preview)



Genuinely idk how it is gonna turn out but I will try to containerize with `macos-latest` runner and will update you. I picked the below as they target pretty much all machines and do tradeoffs whether it is feasible and worth it to replace `ubuntu-22.04-arm` with `macos-latest`.

```bash
      matrix:
        config:
            - { name: 'cpu', runner: 'ubuntu-22.04',     base_image: 'ubuntu:22.04' }
            - { name: 'cpu', runner: 'ubuntu-22.04-arm', base_image: 'ubuntu:22.04' }
            - { name: 'gpu', runner: 'ubuntu-22.04',     base_image: 'nvcr.io/nvidia/nvhpc:23.11-devel-cuda12.3-ubuntu22.04' }
            - { name: 'gpu', runner: 'ubuntu-22.04-arm', base_image: 'nvcr.io/nvidia/nvhpc:23.11-devel-cuda12.3-ubuntu22.04' }
```

sbryngelson commented on docs/documentation/docker.md:
```diff
@@ -0,0 +1,182 @@
+## Tag Details
+### Base Images
+- CPU images (v4.3.0-latest releases) are built on **Ubuntu 22.04**.
+- GPU images (v4.3.0-latest releases) are built on **NVHPC SDK 23.11 (CUDA 12.3) & Ubuntu 22.04**.
```
this names the OS and a compiler

sbryngelson commented on README.md:
```diff
@@ -84,8 +84,11 @@ And here is a high amplitude acoustic wave reflecting and emerging through a cir
+For a quick start, open a GitHub Codespace to load a pre-configured docker container to get familiar with MFC commands. Click <kbd> <> Code</kbd> (green button at top right) → <kbd>Codespaces</kbd> (right tab) → <kbd>+</kbd> (create a codespace).
 
-You can navigate [to this webpage](https://mflowcode.github.io/documentation/md_getting-started.html) to get started using MFC!
+****Notes:**** Codespaces is a free service with a monthly quota of compute time and storage usage. It is commended for testing commands, troubleshooting, and running simple case files without the need to install dependencies and build MFC on your device. Remember to save any important files locally before closing your codespace. To learn more, read through [how docker & containers work](https:// ...
```
should it be Docker or docker?

sbryngelson commented on docs/documentation/getting-started.md:
````diff
@@ -145,6 +145,16 @@ Examples:
+```bash
+docker run -it --rm --entrypoint bash sbryngelson/mfc:latest-cpu
+```
+Please refer to the [Docker](https://mflowcode.github.io/documentation/docker.html) document for more information.
````
should it be called Docker or `Containers`

sbryngelson commented on docs/documentation/getting-started.md:
```diff
@@ -145,6 +145,16 @@ Examples:
 - Build MFC's `simulation` code in Debug mode with MPI and GPU support: `./mfc.sh build --debug --gpu -t simulation`.
 
+## Using Containers
+As an alternative to building MFC from scratch, use containers to quickly access pre-built MFC with all its dependencies.
```
grammar check here

sbryngelson commented on docs/documentation/docker.md:
````diff
@@ -0,0 +1,182 @@
+docker pull sbryngelson/mfc:latest-cpu
+docker save -o mfc:latest-cpu.tar sbryngelson/mfc:latest-cpu
+```
+On the target machine,
````
i don't get why i would do this

sbryngelson commented on docs/documentation/docker.md:
````diff
@@ -0,0 +1,182 @@
+
+Docker by default selects the compatible architecture when pulling and running a container. However, you can manually specify your platform (i.e., `linux/amd64` for most devices or `linux/arm64` for Apple Silicon).
+```bash
+docker run -it --rm --entrypoint bash --platform linux/amd64 sbryngelson/mfc:latest-cpu
````
on Apple Silicon `docker run -it --rm --entrypoint bash sbryngelson/mfc:latest-cpu` works fine for me

Malmahrouqi3 commented on docs/documentation/docker.md:
```diff
@@ -0,0 +1,182 @@
+- Install Docker on [Mac](https://docs.docker.com/desktop/setup/install/mac-install/), [Windows](https://docs.docker.com/desktop/setup/install/windows-install/), or [Linux](https://docs.docker.com/desktop/setup/install/linux/).
+
+Via Docker Desktop GUI,
+- Search for [sbryngelson/mfc](https://hub.docker.com/r/sbryngelson/mfc) repository where all the MFC images are stored then pull a release tag (e.g., `latest-cpu`).
```
The webpage UI & Desktop GUI are not utterly confusing tbh. I would rather keep it pointing to the main page for now.

Malmahrouqi3 commented on docs/documentation/docker.md:
````diff
@@ -0,0 +1,182 @@
+docker run -it --rm --entrypoint bash sbryngelson/mfc:latest-cpu
+```
+<br>
+
````
Good note, the intent is to show detailed guide of docker and how to interact with it while not being isolated from other pages, but yeah sure either re-direct or put in a command or so.

Malmahrouqi3 commented on docs/documentation/docker.md:
````diff
@@ -0,0 +1,182 @@
+
+
+### Slurm Job
+```bash
````
It is just a simple slurm job template. Nothing is universal.

For apptainer/singularity it depends on what cluster you are using - "Consult your HPC User Guide" type thing. PACE, PSC, NCSA, etc. 's wikis give their own instructions on how to properly use containers.

They might not be on a login node, but typically are pre-loaded on a compute node.

Malmahrouqi3 commented on docs/documentation/docker.md:
````diff
@@ -0,0 +1,182 @@
+
+### **Interactive Shell**
+```bash
+apptainer shell --nv --fakeroot --writable-tmpfs --bind "$PWD":/mnt  docker://sbryngelson/mfc:latest-gpu
````
NVIDIA GPUs have been the main focus for HPC containers.
cpu cases on clusters - that would be the same commands but with the `latest-cpu` container and without the `--nv` flag.
amd gpu cases - just clone & build MFCs, not worth the hassle. Those amd gpus would be obsolete. Unless you are willing to either 1) containerize on a machine with amd compilers and root access or 2) pull a standard image, re-build MFC with the HPC host's compilers and re-package the container.

Containers are just VMs that utilize host's resources and not necessarily with bi-directional access on HPC - meaning re-packaging is a different process from containerizing with its own hurdles.

Malmahrouqi3 commented:
I made a google colab with Getting-Started and Docker pages' content.
I asked someone in a call with no code or MFC background to attempt the commented sentences with no external help.
Pasted below is his words on each quest.

```python
# Install Linux and run all commands there. 
# Note: use this cell to paste your code. 
# Use docker to run a 2D_shockbubble case on v4.9.9 CPU release.

First press the windows key, search for the Docker app. Open the Docker app and go to the search bar. In the search bar, type the following search as follows: " sbryngelson/mfc" Upon the results that are shown, 
hover over the first option shown, then pan over to the the tag menu. Scroll down until you have found the requested version, "v4.9.9cpu". Confirm your choice by clicking on "v4.9.9cpu". 
Click run to execute what you have searched. Inspect the file as it's running by checking on the terminal, located in  the "Execution" tab. You can use this to run any command in the terminal as the program is running. 
Next, in the Terminal, where the Program is running, paste the following command: "./mfc.sh run examples/2D_shockbubble/case.py -n 2".

# Use docker to run all tests on on v5.0.5 GPU release.

Copy the following command: "docker run -it --rm --gpus all --entrypoint bash sbryngelson/mfc:latest-gpu". Replace the "latest" tab with the requested GPU version, which in this instance, would be "v5.0.5". 
Ensure the V is lowercase, as the command is case sensitive. Proceed to the Terminal in the Docker app, located in the lower right hand corner of the screen. 
Paste the command, ensuring to replace "latest" with the specified GPU version, "v5.0.5". Excecute (run) the command. Copy the following command: "./mfc.sh test -j 8" into the terminal of Docker, and run it.
```
